File: phase-2/frontend/lib/types/category.ts
/**
 * TypeScript types for Category entity.
 *
 * Matches backend Category SQLModel schema from:
 * phase-2/backend/src/models/category.py
 *
 * @see specs/001-fullstack-todo-web/spec.md - Data Structures
 */

/**
 * Category entity - full category representation.
 *
 * Matches CategoryRead schema from backend.
 */
export interface Category {
  /** Unique identifier (UUID4) */
  id: string;
  /** Owner user ID (UUID4, null for system categories) */
  user_id: string | null;
  /** Category name (1-50 characters, unique per user) */
  name: string;
  /** Optional hex color code (e.g., "#FF5733") */
  color: string | null;
  /** Whether this is a system-defined category */
  is_system: boolean;
  /** Creation timestamp (ISO 8601) */
  created_at: string;
}

/**
 * Category creation payload.
 *
 * Matches CategoryCreate schema from backend.
 */
export interface CategoryCreate {
  /** Category name (required, 1-50 characters) */
  name: string;
  /** Optional hex color code */
  color?: string | null;
}

/**
 * Category update payload.
 *
 * Matches CategoryUpdate schema from backend.
 * All fields are optional - only provided fields are updated.
 */
export interface CategoryUpdate {
  /** Updated name */
  name?: string;
  /** Updated color */
  color?: string | null;
}

/**
 * System category names (pre-seeded, read-only).
 */
export const SYSTEM_CATEGORIES = [
  "Work",
  "Personal",
  "Shopping",
  "Health",
  "Finance",
] as const;

export type SystemCategoryName = (typeof SYSTEM_CATEGORIES)[number];

/**
 * Category list response.
 */
export interface CategoryListResponse {
  /** List of categories */
  categories: Category[];
  /** Total count */
  total: number;
}

/**
 * Category with task count.
 *
 * Used for displaying categories with usage statistics.
 */
export interface CategoryWithCount extends Category {
  /** Number of tasks in this category */
  task_count: number;
}

File: phase-3/backend/src/models/category.py
"""Category SQLModel - Task organization by category."""

from datetime import datetime
from typing import Optional
from uuid import UUID, uuid4
import sqlalchemy as sa
from sqlmodel import Field, SQLModel, Column


class Category(SQLModel, table=True):
    """Category entity for task classification."""

    __tablename__ = "categories"

    id: UUID = Field(default_factory=uuid4, primary_key=True, nullable=False)
    user_id: Optional[str] = Field(default=None, sa_column=Column(sa.VARCHAR(64), nullable=True, index=True))
    name: str = Field(max_length=100, nullable=False)
    is_system: bool = Field(default=False, nullable=False, index=True)
    color: Optional[str] = Field(default=None, max_length=7)  # Hex color code
    created_at: datetime = Field(default_factory=datetime.utcnow, nullable=False)


class CategoryCreate(SQLModel):
    """Schema for creating a custom category."""
    name: str
    color: Optional[str] = None


class CategoryRead(SQLModel):
    """Schema for category responses."""
    id: UUID
    user_id: Optional[str]
    name: str
    is_system: bool
    color: Optional[str]
    created_at: datetime


class CategoryUpdate(SQLModel):
    """Schema for updating a category."""
    name: Optional[str] = None
    color: Optional[str] = None
